fraud/service.go

```diff
@@ -0,0 +1,192 @@
+	}
+
+	return topic, ok, nil
+}
+
+// publish allows to publish Fraud Proofs to the network
+func (t *topics) publish(ctx context.Context, data []byte, proofType ProofType) error {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+
+	if topic, ok := t.pubSubTopics[proofType]; ok {
+		return topic.Publish(ctx, data)
+	}
+
+	return errors.New("topic is not found")
+}
+
+// service is propagating and validating Fraud Proofs
+// service implements Service interface.
+type service struct {
+	topics       *topics
+	unmarshalers map[ProofType]ProofUnmarshaler
+	getter       headerFetcher
+	mu           sync.RWMutex
+}
+
+func NewService(p *pubsub.PubSub, getter headerFetcher) Service {
+	return &service{
+		topics: &topics{
+			pubsub:       p,
+			pubSubTopics: make(map[ProofType]*pubsub.Topic),
+		},
+		unmarshalers: make(map[ProofType]ProofUnmarshaler),
+		getter:       getter,
+	}
+}
+
+func (f *service) Start(context.Context) error {
+	log.Info("fraud service is starting...")
+
+	err := f.RegisterUnmarshaler(BadEncoding, UnmarshalBEFP)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (f *service) Stop(context.Context) error {
+	log.Info("fraud service is stopping...")
+	return nil
+}
+
+func (f *service) Subscribe(proofType ProofType) (Subscription, error) {
+	// TODO: @vgonkivs check if fraud proof is in fraud store, then return with error
+	f.mu.RLock()
+	u, ok := f.unmarshalers[proofType]
+	f.mu.RUnlock()
+	if !ok {
+		return nil, errors.New("unmarshaler is not registered")
+	}
+
+	t, wasjoined, err := f.topics.getTopic(proofType)
+	if err != nil {
+		return nil, err
+	}
+	// if topic was joined for the first time then we should register a validator for it
+	if !wasjoined {
+		// add internal validation to topic inside libp2p for provided ProofType
+		if err = f.topics.pubsub.RegisterTopicValidator(
+			getSubTopic(proofType),
+			func(ctx context.Context, _ peer.ID, msg *pubsub.Message) pubsub.ValidationResult {
+				return f.processIncoming(ctx, proofType, msg.Data)
+			},
+		); err != nil {
+			return nil, err
+		}
+	}
+
+	return newSubscription(t, u)
+}
+
+func (f *service) RegisterUnmarshaler(proofType ProofType, u ProofUnmarshaler) error {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+
+	if _, ok := f.unmarshalers[proofType]; ok {
+		return errors.New("unmarshaler is registered")
+	}
+	f.unmarshalers[proofType] = u
+
+	return nil
+}
+
+func (f *service) UnregisterUnmarshaler(proofType ProofType) error {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+
+	if _, ok := f.unmarshalers[proofType]; !ok {
+		return errors.New("unmarshaler is not registered")
+	}
+	delete(f.unmarshalers, proofType)
+
+	return nil
+
+}
+
+func (f *service) Broadcast(ctx context.Context, p Proof) error {
+	bin, err := p.MarshalBinary()
+	if err != nil {
+		return err
+	}
+
+	return f.topics.publish(ctx, bin, p.Type())
+}
+
+func (f *service) processIncoming(ctx context.Context, proofType ProofType, data []byte) pubsub.ValidationResult {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
```

Comment:
The same issue with defer here. `proof.Validate` can take some time, same for the `f.getter`. Taking the lock all that time is not really a problem for the case of fraud proofs, but this is more like pointing out that this issue can happen in other hotter code paths you write. This can easily become a contention problem if this was e.g. a msg processing.

Reply:
unlocking a mutex after picking an unmarshaler from map